File: terminal_ticker/market_data/candle_cache.py
"""文件用途：数据源层，用 SQLite 缓存标准化 K 线并支持增量拉取。"""
from __future__ import annotations

from dataclasses import dataclass
import os
from pathlib import Path
import sqlite3
import time
from typing import Callable

from ..config import CacheConfig
from ..domain.price_action import Candle

DEFAULT_CACHE_SUBDIR = "terminal-ticker"
DEFAULT_CACHE_FILENAME = "candles.sqlite3"
INTERVAL_SECONDS = {
    "1m": 60,
    "3m": 3 * 60,
    "5m": 5 * 60,
    "15m": 15 * 60,
    "30m": 30 * 60,
    "1H": 60 * 60,
    "4H": 4 * 60 * 60,
    "6H": 6 * 60 * 60,
    "12H": 12 * 60 * 60,
    "1D": 24 * 60 * 60,
    "3D": 3 * 24 * 60 * 60,
    "1W": 7 * 24 * 60 * 60,
    "1M": 31 * 24 * 60 * 60,
}


@dataclass(frozen=True)
class CandleFetchPlan:
    """说明：封装读取缓存后下一次 provider 请求的计划。"""
    after_open_time_ms: int | None
    limit: int


def default_cache_path() -> Path:
    """说明：返回平台本地默认 SQLite 缓存路径。"""
    base = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
    return base / DEFAULT_CACHE_SUBDIR / DEFAULT_CACHE_FILENAME


class CandleCache:
    """说明：封装按标的和周期存取 K 线的 SQLite 缓存。"""

    def __init__(self, path: str | Path, *, retention_seconds: int = 86_400) -> None:
        """说明：初始化当前对象的运行状态。"""
        self.path = Path(path).expanduser()
        self.retention_seconds = retention_seconds

    @classmethod
    def from_config(cls, config: CacheConfig) -> "CandleCache":
        """说明：根据应用配置创建缓存实例。"""
        return cls(
            config.path or default_cache_path(),
            retention_seconds=config.candle_retention_seconds,
        )

    def _connect(self) -> sqlite3.Connection:
        """说明：打开连接并确保底层资源已经初始化。"""
        self.path.parent.mkdir(parents=True, exist_ok=True)
        connection = sqlite3.connect(self.path)
        connection.execute(
            """
            CREATE TABLE IF NOT EXISTS candles (
                symbol_key TEXT NOT NULL,
                interval TEXT NOT NULL,
                open_time_ms INTEGER NOT NULL,
                open REAL NOT NULL,
                high REAL NOT NULL,
                low REAL NOT NULL,
                close REAL NOT NULL,
                volume REAL NOT NULL,
                fetched_at_ms INTEGER NOT NULL,
                PRIMARY KEY (symbol_key, interval, open_time_ms)
            )
            """
        )
        connection.execute(
            """
            CREATE INDEX IF NOT EXISTS idx_candles_lookup
            ON candles (symbol_key, interval, open_time_ms)
            """
        )
        return connection

    def prune(
        self,
        *,
        now_ms: int | None = None,
        retention_seconds: int | None = None,
    ) -> int:
        """说明：删除超过保留期的 K 线缓存。"""
        current_ms = _now_ms() if now_ms is None else now_ms
        cutoff_ms = current_ms - self._effective_retention(retention_seconds) * 1000
        with self._connect() as connection:
            cursor = connection.execute(
                "DELETE FROM candles WHERE open_time_ms < ?",
                (cutoff_ms,),
            )
            return cursor.rowcount

    def upsert(
        self,
        candles: tuple[Candle, ...],
        *,
        interval: str,
        fetched_at_ms: int | None = None,
    ) -> int:
        """说明：把 K 线写入缓存，已存在时更新。"""
        if not candles:
            return 0
        current_ms = _now_ms() if fetched_at_ms is None else fetched_at_ms
        rows = [
            (
                candle.symbol_key,
                interval,
                candle.open_time_ms,
                candle.open,
                candle.high,
                candle.low,
                candle.close,
                candle.volume,
                current_ms,
            )
            for candle in candles
        ]
        with self._connect() as connection:
            connection.executemany(
                """
                INSERT INTO candles (
                    symbol_key, interval, open_time_ms, open, high, low, close, volume,
                    fetched_at_ms
                )
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(symbol_key, interval, open_time_ms) DO UPDATE SET
                    open = excluded.open,
                    high = excluded.high,
                    low = excluded.low,
                    close = excluded.close,
                    volume = excluded.volume,
                    fetched_at_ms = excluded.fetched_at_ms
                """,
                rows,
            )
        return len(rows)

    def latest_open_time_ms(
        self,
        symbol_key: str,
        interval: str,
        *,
        now_ms: int | None = None,
        retention_seconds: int | None = None,
    ) -> int | None:
        """说明：读取某个标的和周期最新的缓存 K 线开盘时间。"""
        current_ms = _now_ms() if now_ms is None else now_ms
        cutoff_ms = current_ms - self._effective_retention(retention_seconds) * 1000
        with self._connect() as connection:
            row = connection.execute(
                """
                SELECT MAX(open_time_ms)
                FROM candles
                WHERE symbol_key = ? AND interval = ? AND open_time_ms >= ?
                """,
                (symbol_key, interval, cutoff_ms),
            ).fetchone()
        value = row[0] if row else None
        return int(value) if value is not None else None

    def recent(
        self,
        symbol_key: str,
        interval: str,
        *,
        limit: int,
        now_ms: int | None = None,
        retention_seconds: int | None = None,
    ) -> tuple[Candle, ...]:
        """说明：读取某个标的和周期最近的缓存 K 线。"""
        current_ms = _now_ms() if now_ms is None else now_ms
        cutoff_ms = current_ms - self._effective_retention(retention_seconds) * 1000
        with self._connect() as connection:
            rows = connection.execute(
                """
                SELECT symbol_key, open_time_ms, open, high, low, close, volume
                FROM candles
                WHERE symbol_key = ? AND interval = ? AND open_time_ms >= ?
                ORDER BY open_time_ms DESC
                LIMIT ?
                """,
                (symbol_key, interval, cutoff_ms, limit),
            ).fetchall()
        candles = tuple(
            Candle(
                symbol_key=str(row[0]),
                open_time_ms=int(row[1]),
                open=float(row[2]),
                high=float(row[3]),
                low=float(row[4]),
                close=float(row[5]),
                volume=float(row[6]),
            )
            for row in reversed(rows)
        )
        return candles

    def _effective_retention(self, requested_seconds: int | None) -> int:
        """说明：返回本次读取所需的缓存保留秒数。"""
        if requested_seconds is None:
            return self.retention_seconds
        return max(self.retention_seconds, requested_seconds)


def cached_fetch_candles(
    *,
    cache: CandleCache,
    symbol_key: str,
    interval: str,
    limit: int,
    fetcher: Callable[..., tuple[Candle, ...]],
    now_ms: int | None = None,
    minimum_retention_seconds: int | None = None,
) -> tuple[Candle, ...]:
    """说明：通过 SQLite 缓存拉取 K 线，并只向 provider 请求缺失区间。"""
    current_ms = _now_ms() if now_ms is None else now_ms
    cache.prune(now_ms=current_ms, retention_seconds=minimum_retention_seconds)
    latest_open_ms = cache.latest_open_time_ms(
        symbol_key,
        interval,
        now_ms=current_ms,
        retention_seconds=minimum_retention_seconds,
    )
    plan = _fetch_plan(latest_open_ms, interval=interval, limit=limit, now_ms=current_ms)
    try:
        fetched = fetcher(
            interval=interval,
            limit=plan.limit,
            after_open_time_ms=plan.after_open_time_ms,
        )
    except Exception:
        cached = cache.recent(
            symbol_key,
            interval,
            limit=limit,
            now_ms=current_ms,
            retention_seconds=minimum_retention_seconds,
        )
        if cached:
            return cached
        raise

    if fetched:
        cache.upsert(fetched, interval=interval, fetched_at_ms=current_ms)
    cached = cache.recent(
        symbol_key,
        interval,
        limit=limit,
        now_ms=current_ms,
        retention_seconds=minimum_retention_seconds,
    )
    return cached or fetched[-limit:]


def _fetch_plan(
    latest_open_ms: int | None,
    *,
    interval: str,
    limit: int,
    now_ms: int,
) -> CandleFetchPlan:
    """说明：根据缓存最新时间计算下一次 provider 请求计划。"""
    if latest_open_ms is None:
        return CandleFetchPlan(after_open_time_ms=None, limit=limit)
    interval_ms = INTERVAL_SECONDS.get(interval, 60) * 1000
    missing = max(1, ((now_ms - latest_open_ms) // interval_ms) + 2)
    if missing > 1000:
        return CandleFetchPlan(after_open_time_ms=None, limit=min(max(limit, 1000), missing))
    overlapped_after_open_ms = max(0, latest_open_ms - interval_ms)
    return CandleFetchPlan(
        after_open_time_ms=overlapped_after_open_ms,
        limit=max(limit, int(missing)),
    )


def _now_ms() -> int:
    """说明：返回当前 Unix 毫秒时间戳。"""
    return int(time.time() * 1000)

File: tradex/market_data/candle-cache.ts
import Database from "better-sqlite3";
import path from "node:path";
import { CacheConfig } from "../config/index.js";
import { BaseStore, defaultCacheDir, nowMs } from "../db.js";
import { Candle } from "../domain/price-action.js";

export const DEFAULT_CACHE_FILENAME = "candles.sqlite3";
const INTERVAL_SECONDS: Record<string, number> = {
  "1m": 60,
  "3m": 180,
  "5m": 300,
  "15m": 900,
  "30m": 1800,
  "1H": 3600,
  "4H": 14_400,
  "6H": 21_600,
  "12H": 43_200,
  "1D": 86_400,
  "3D": 259_200,
  "1W": 604_800,
  "1M": 2_678_400,
};
const DIRECT_WINDOW_CANDLE_LIMIT = 1000;

export interface CandleFetchPlan {
  afterOpenTimeMs: number | null;
  limit: number;
}

export function defaultCachePath(): string {
  return path.join(defaultCacheDir(), DEFAULT_CACHE_FILENAME);
}

export function retentionSecondsForWindow(interval: string, limit: number): number {
  return (INTERVAL_SECONDS[interval] ?? 60) * Math.max(1, limit);
}

export class CandleCache extends BaseStore {
  readonly retentionSeconds: number;

  constructor(dbPath: string, retentionSeconds = 86_400) {
    super(dbPath);
    this.retentionSeconds = retentionSeconds;
  }

  static fromConfig(config: CacheConfig): CandleCache {
    return new CandleCache(config.path || defaultCachePath(), config.candleRetentionSeconds);
  }

  protected override initSchema(conn: Database.Database): void {
    conn.exec(`
      CREATE TABLE IF NOT EXISTS candles (
        symbol_key TEXT NOT NULL,
        interval TEXT NOT NULL,
        open_time_ms INTEGER NOT NULL,
        open REAL NOT NULL,
        high REAL NOT NULL,
        low REAL NOT NULL,
        close REAL NOT NULL,
        volume REAL NOT NULL,
        fetched_at_ms INTEGER NOT NULL,
        PRIMARY KEY (symbol_key, interval, open_time_ms)
      );
      CREATE INDEX IF NOT EXISTS idx_candles_lookup
      ON candles (symbol_key, interval, open_time_ms);
    `);
  }

  prune(input: { nowMs?: number; retentionSeconds?: number | null; symbolKey?: string | null; interval?: string | null } = {}): number {
    const currentMs = input.nowMs ?? nowMs();
    const cutoffMs = currentMs - this.effectiveRetention(input.retentionSeconds ?? null) * 1000;
    const clauses = ["open_time_ms < ?"];
    const params: Array<string | number> = [cutoffMs];
    if (input.symbolKey) {
      clauses.push("symbol_key = ?");
      params.push(input.symbolKey);
    }
    if (input.interval) {
      clauses.push("interval = ?");
      params.push(input.interval);
    }
    return this.getConn().prepare(`DELETE FROM candles WHERE ${clauses.join(" AND ")}`).run(...params).changes;
  }

  upsert(candles: readonly Candle[], input: { interval: string; fetchedAtMs?: number | null }): number {
    if (candles.length === 0) return 0;
    const fetchedAtMs = input.fetchedAtMs ?? nowMs();
    const stmt = this.getConn().prepare(`
      INSERT INTO candles (symbol_key, interval, open_time_ms, open, high, low, close, volume, fetched_at_ms)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
      ON CONFLICT(symbol_key, interval, open_time_ms) DO UPDATE SET
        open = excluded.open,
        high = excluded.high,
        low = excluded.low,
        close = excluded.close,
        volume = excluded.volume,
        fetched_at_ms = excluded.fetched_at_ms
    `);
    const tx = this.getConn().transaction((rows: readonly Candle[]) => {
      for (const candle of rows) {
        stmt.run(candle.symbolKey, input.interval, candle.openTimeMs, candle.open, candle.high, candle.low, candle.close, candle.volume, fetchedAtMs);
      }
    });
    tx(candles);
    return candles.length;
  }

  latestFetchedAtMs(symbolKey: string, interval: string, input: { nowMs?: number; retentionSeconds?: number | null } = {}): number | null {
    const currentMs = input.nowMs ?? nowMs();
    const cutoffMs = currentMs - this.effectiveRetention(input.retentionSeconds ?? null) * 1000;
    const row = this.getConn()
      .prepare(
        `SELECT fetched_at_ms AS fetchedAtMs
         FROM candles
         WHERE symbol_key = ? AND interval = ? AND open_time_ms >= ?
         ORDER BY open_time_ms DESC
         LIMIT 1`,
      )
      .get(symbolKey, interval, cutoffMs) as { fetchedAtMs?: number } | undefined;
    return row?.fetchedAtMs ?? null;
  }

  recent(symbolKey: string, interval: string, input: { limit: number; nowMs?: number; retentionSeconds?: number | null }): Candle[] {
    const currentMs = input.nowMs ?? nowMs();
    const cutoffMs = currentMs - this.effectiveRetention(input.retentionSeconds ?? null) * 1000;
    const rows = this.getConn()
      .prepare(
        `SELECT symbol_key AS symbolKey, open_time_ms AS openTimeMs, open, high, low, close, volume
         FROM candles
         WHERE symbol_key = ? AND interval = ? AND open_time_ms >= ?
         ORDER BY open_time_ms DESC
         LIMIT ?`,
      )
      .all(symbolKey, interval, cutoffMs, input.limit) as Array<{
      symbolKey: string;
      openTimeMs: number;
      open: number;
      high: number;
      low: number;
      close: number;
      volume: number;
    }>;
    return rows
      .reverse()
      .map(
        (row) =>
          new Candle({
            symbolKey: row.symbolKey,
            openTimeMs: row.openTimeMs,
            open: row.open,
            high: row.high,
            low: row.low,
            close: row.close,
            volume: row.volume,
          }),
      );
  }

  private effectiveRetention(requestedSeconds: number | null): number {
    return requestedSeconds === null ? this.retentionSeconds : Math.max(this.retentionSeconds, requestedSeconds);
  }
}

export async function cachedFetchCandles(input: {
  cache: CandleCache;
  symbolKey: string;
  interval: string;
  limit: number;
  fetcher: (args: { interval: string; limit: number; afterOpenTimeMs?: number | null }) => Promise<Candle[]>;
  nowMs?: number;
  minimumRetentionSeconds?: number | null;
  maxCacheAgeSeconds?: number | null;
}): Promise<Candle[]> {
  const currentMs = input.nowMs ?? nowMs();
  input.cache.prune({
    nowMs: currentMs,
    retentionSeconds: input.minimumRetentionSeconds ?? null,
    symbolKey: input.symbolKey,
    interval: input.interval,
  });
  let cached = input.cache.recent(input.symbolKey, input.interval, {
    limit: input.limit,
    nowMs: currentMs,
    retentionSeconds: input.minimumRetentionSeconds ?? null,
  });
  if (input.maxCacheAgeSeconds !== undefined && input.maxCacheAgeSeconds !== null) {
    const latestFetchedAtMs = input.cache.latestFetchedAtMs(input.symbolKey, input.interval, {
      nowMs: currentMs,
      retentionSeconds: input.minimumRetentionSeconds ?? null,
    });
    if (cached.length >= input.limit && latestFetchedAtMs !== null && currentMs - latestFetchedAtMs <= input.maxCacheAgeSeconds * 1000) {
      return cached;
    }
  }
  const plan = fetchPlan(cached.at(-1)?.openTimeMs ?? null, cached.length, input.interval, input.limit, currentMs);
  try {
    const fetched = await input.fetcher({ interval: input.interval, limit: plan.limit, afterOpenTimeMs: plan.afterOpenTimeMs });
    if (fetched.length > 0) input.cache.upsert(fetched, { interval: input.interval, fetchedAtMs: currentMs });
    cached = input.cache.recent(input.symbolKey, input.interval, {
      limit: input.limit,
      nowMs: currentMs,
      retentionSeconds: input.minimumRetentionSeconds ?? null,
    });
    return cached.length > 0 ? cached : fetched.slice(-input.limit);
  } catch (error) {
    cached = input.cache.recent(input.symbolKey, input.interval, {
      limit: input.limit,
      nowMs: currentMs,
      retentionSeconds: input.minimumRetentionSeconds ?? null,
    });
    if (cached.length > 0) return cached;
    throw error;
  }
}

function fetchPlan(latestOpenMs: number | null, cachedCount: number, interval: string, limit: number, currentMs: number): CandleFetchPlan {
  if (latestOpenMs === null || cachedCount < limit) return { afterOpenTimeMs: null, limit: Math.max(limit, DIRECT_WINDOW_CANDLE_LIMIT) };
  const intervalMs = (INTERVAL_SECONDS[interval] ?? 60) * 1000;
  const missing = Math.max(1, Math.floor((currentMs - latestOpenMs) / intervalMs) + 2);
  if (missing > DIRECT_WINDOW_CANDLE_LIMIT) return { afterOpenTimeMs: null, limit: Math.max(limit, DIRECT_WINDOW_CANDLE_LIMIT) };
  return { afterOpenTimeMs: Math.max(0, latestOpenMs - intervalMs), limit: missing };
}
